analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyLoggingCalls.cs

```diff
@@ -0,0 +1,217 @@
+                var invocations = loggingCategory.Value;
+                if (invocations.Count > treshold)
+                {
+                    var primaryLocation = invocations[0].GetLocation();
+                    var secondaryLocations = invocations.Skip(1).Select(x => x.GetLocation());
+                    context.ReportIssue(Diagnostic.Create(Rule, primaryLocation, secondaryLocations, loggingCategory.Key, invocations.Count, treshold));
+                }
+            }
+        }
+    }
+
+    private int Treshold(string category)
+    {
+        var threshold = category switch
+        {
+            CategoryNames.Debug => DebugThreshold,
+            CategoryNames.Error => ErrorThreshold,
+            CategoryNames.Warning => WarningThreshold,
+            _ => InformationThreshold
+        };
+        return Math.Max(threshold, 0);
+    }
+
+    private static bool IsBlockNodeSupported(SyntaxNode node) =>
+        node.Kind() == SyntaxKind.Block
+        || (node.Kind() == SyntaxKind.CompilationUnit && node.ChildNodes().OfType<GlobalStatementSyntax>().Any());  // for top-level statements
+
+    private sealed class LoggingCallCollector : SafeCSharpSyntaxWalker
+    {
+        private readonly SemanticModel semanticModel;
+        private SyntaxNode currentBlock;
+
+        public Dictionary<string, List<InvocationExpressionSyntax>> LoggingInvocations { get; } = [];
+
+        public LoggingCallCollector(SemanticModel semanticModel, SyntaxNode currentBlock)
+        {
+            this.semanticModel = semanticModel;
+            this.currentBlock = currentBlock;
+        }
+
+        public override void VisitBlock(BlockSyntax node)
+        {
+            if (currentBlock == node) // Don't visit nested blocks, that will be done by additional LoggingCallCollector instances.
+            {
+                currentBlock = node;
+                base.VisitBlock(node);
+            }
+        }
+
+        public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+        {
+            if (IsLoggerMethod(node.GetName())
+                && semanticModel.GetSymbolInfo(node).Symbol is IMethodSymbol methodSymbol
+                && methodSymbol.ContainingType.DerivesFromAny(LoggerTypes)
+                && LoggingCategoryName(node, methodSymbol) is { } loggingCategory)
+            {
+                if (LoggingInvocations.TryGetValue(loggingCategory, out var invocationList))
+                {
+                    invocationList.Add(node);
+                }
+                else
+                {
+                    LoggingInvocations.Add(loggingCategory, [node]);
+                }
+            }
+        }
+
+        public override void VisitIfStatement(IfStatementSyntax node)
+        {
+            // Skip syntax node to avoid FPs
+        }
+
+        public override void VisitParenthesizedLambdaExpression(ParenthesizedLambdaExpressionSyntax node)
+        {
+            // Skip syntax node to avoid FPs
```

Comment:
Lambdas have their own `Block`s, right?

E.g.

![image](https://github.com/SonarSource/sonar-dotnet/assets/115458417/2f728a99-8be7-448d-9d29-1a542151cda0)


![image](https://github.com/SonarSource/sonar-dotnet/assets/115458417/d8ee9767-7b07-4693-a124-4533b5004987)


What FP are you refering to?

Reply:
[The body of a lambda can be a single instruction](https://sharplab.io/#v2:C4LghgzsA0AmIGoA+ABATARgLACgUGYACdQgYUIG9dCbiiUAWQgWQAoBKS62nlDNQmADGwAJYB7AHaEAvIQ6yAfMQwBOVgCIAEgFMANnvGEAhBvYBubjQC+ua0A=), it doesn't have to be a block.
[Here's a potential FP](https://github.com/SonarSource/sonar-dotnet/blob/18e46259cfe3eb14695507014b26fc2674669020/analyzers/tests/SonarAnalyzer.Test/TestCases/TooManyLoggingCalls.cs#L159).